lib/Frontend/Frontend.cpp

```diff
@@ -190,107 +192,111 @@ bool CompilerInstance::setUpInputs(Optional<unsigned> codeCompletionBufferID) {
-  }
-  return setUpForFile(input.file(), input.isPrimary());
-}
-void CompilerInstance::setUpForBuffer(llvm::MemoryBuffer *buffer,
-                                      bool isPrimary) {
-  auto copy = llvm::MemoryBuffer::getMemBufferCopy(
-      buffer->getBuffer(), buffer->getBufferIdentifier());
-  if (serialization::isSerializedAST(copy->getBuffer())) {
-    PartialModules.push_back({std::move(copy), nullptr});
-  } else {
-    unsigned bufferID = SourceMgr.addNewSourceBuffer(std::move(copy));
-    InputSourceCodeBufferIDs.push_back(bufferID);
 
-    if (isInSILMode()) {
-      assert(MainBufferID == NO_SUCH_BUFFER && "re-setting MainBufferID");
-      MainBufferID = bufferID;
-    }
+  if (isInSILMode() ||
+      (input.buffer() == nullptr && isInputSwift() &&
+       llvm::sys::path::filename(input.file()) == "main.swift")) {
+    assert(MainBufferID == NO_SUCH_BUFFER && "re-setting MainBufferID");
+    MainBufferID = *bufferID;
+  }
 
-    if (isPrimary) {
-      assert(PrimaryBufferID == NO_SUCH_BUFFER && "re-setting PrimaryBufferID");
-      PrimaryBufferID = bufferID;
-    }
+  if (input.isPrimary()) {
+    assert(PrimaryBufferID == NO_SUCH_BUFFER && "re-setting PrimaryBufferID");
+    PrimaryBufferID = *bufferID;
   }
+  return false;
 }
 
-bool CompilerInstance::setUpForFile(StringRef fileName, bool isPrimary) {
-  if (fileName.empty())
-    return false;
-  // FIXME: Working with filenames is fragile, maybe use the real path
-  // or have some kind of FileManager.
-  using namespace llvm::sys::path;
-  if (Optional<unsigned> existingBufferID =
-          SourceMgr.getIDForBufferIdentifier(fileName)) {
-    if (isInSILMode() ||
-        (isInputSwift() && filename(fileName) == "main.swift")) {
-      assert(MainBufferID == NO_SUCH_BUFFER && "re-setting MainBufferID");
-      MainBufferID = existingBufferID.getValue();
-    }
-
-    if (isPrimary) {
-      assert(PrimaryBufferID == NO_SUCH_BUFFER && "re-setting PrimaryBufferID");
-      PrimaryBufferID = existingBufferID.getValue();
+Optional<unsigned> CompilerInstance::getRecordedBufferID(const InputFile &input,
+                                                         bool &failed) {
+  if (!input.buffer()) {
+    if (Optional<unsigned> existingBufferID =
+            SourceMgr.getIDForBufferIdentifier(input.file())) {
+      return existingBufferID;
     }
-
-    return false; // replaced by a memory buffer.
   }
+  std::pair<std::unique_ptr<llvm::MemoryBuffer>,
+            std::unique_ptr<llvm::MemoryBuffer>>
+      buffers = getInputBufferAndModuleDocBufferIfPresent(input);
 
-  // Open the input file.
-  using FileOrError = llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>>;
-  FileOrError inputFileOrErr = llvm::MemoryBuffer::getFileOrSTDIN(fileName);
-  if (!inputFileOrErr) {
-    Diagnostics.diagnose(SourceLoc(), diag::error_open_input_file, fileName,
-                         inputFileOrErr.getError().message());
-    return true;
+  if (!buffers.first.get()) {
```

Comment:
`std::unique_ptr` has a conversion to `bool`, so no need for the `get()`.

Reply:
Removed.